Deconstructing `thread::scope` (#120)
ferrous-systems/rust-training: training-slides/src/deconstructing-thread-scope.md

pub struct Scope<'scope, 'env: 'scope> {
    data: Arc<ScopeData>,
    scope: PhantomData<&'scope mut &'scope ()>,
    env: PhantomData<&'env mut &'env ()>,
}
```

## `Scope::spawn`

```rust ignore
impl<'scope, 'env> Scope<'scope, 'env> {
    pub fn spawn<F, T>(&'scope self, f: F) -> ScopedJoinHandle<'scope, T>
    where
        F: FnOnce() -> T + Send + 'scope,
        T: Send + 'scope,
    {
        // ...
    }
}
```

we'll call `f` a **Thread** closure

## Closures

```rust
fn main() {
    let data = vec![1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
    let (left, right) = data.split_at(data.len() / 2);
    let (mut left_sum, mut right_sum) = (0, 0);
    std::thread::scope(|s| { // <- Spawner closure
        s.spawn(|| { // <- Thread closure
            left_sum = left.iter().sum();
        });
        s.spawn(|| { // <- Thread closure
            right_sum = right.iter().sum();
        });
    });
    println!("Total: {}", left_sum + right_sum);
}
```

## Meet lifetime annotations

* `F: for<'scope> FnOnce(&'scope Scope<'scope, 'env>) -> T`
  * `'scope` is a *Higher-Rank Trait Bound* - it describes *all possible* lifetimes that closure can observe
* `fn scope<'env, F, T>(f: F) -> T`

>   * the function observes *some* lifetime `'env`

Reviewer: ```suggestion
  * the function observes *some* lifetime `'env`
```

In the lifetimes deck we tried to distinguish between a lifetime (which is a property of a value, marking the time between its creation and its destruction as observed at compile-time), and a lifetime tag/annotation/specifier which is what we add to references to tag or annotate the lifetime of the value they refer to so we can tag/annotate other references as having a *must-not-outlive* relationship.

## Thread closure

```text
F: FnOnce() -> T + Send + 'scope,
T: Send + 'scope
```

* Data passed to and from the child thread
  * should be `Send`
  * if has references to surrounding data, they should stay valid *at least* for the whole duration of `'scope`
    * `&'scope Scope` in **Spawner** signature allows nested calls to `spawn`!

## Relationship 2

By the time `'scope` is over all calls to `Scope::spawn` are over and all **Thread** closures are completed.

## Practical implications

* **Spawner** closure can finish earlier that **Thread** closures.
  * **Thread** closures can't use upvars from **Spawner** without moving them
* both **Spawner** and all **Thread** closures are completed before the call to `scope` function returns
  * both can take upvars from the code before `scope()` call without moving

## Relationships in action

```rust ignore
fn main() {
    let data = vec![1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
    let (left, right) = data.split_at(data.len() / 2);
    let mut left_sum = 0; // Ok
    std::thread::scope(|s| {
        s.spawn(|| {
            left_sum = left.iter().sum();
        });
        // ERROR: closure may outlive the current function

>         let mut right_sum = 0;

Reviewer: I think I prefer having examples which compile and run, but where the 'bad code' is commented out.

        s.spawn(|| {
            right_sum = right.iter().sum();
        });
    });
    // println!("Total: {}", left_sum + right_sum);
}
```

## Using `join`

```rust
fn main() {
    let data = vec![1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
    let (left, right) = data.split_at(data.len() / 2);
    let mut left_sum = 0;
    let right_sum = std::thread::scope(|s| {
        s.spawn(|| {
            left_sum = left.iter().sum();
        });

        let handle = s.spawn(|| {
            right.iter().sum()
        });
        let right_sum: i32 = handle.join().unwrap();
        right_sum
    });
    println!("Total: {}", left_sum + right_sum);
}
```

> ## How `scope` function waits for all threads to finish?

Reviewer: ```suggestion
## How does `fn scope()` wait for all threads to finish?
```

```rust ignore
pub fn scope<'env, F, T>(f: F) -> T {
    let scope = Scope {
        data: Arc::new(ScopeData {
            num_running_threads: AtomicUsize::new(0),
            main_thread: current(),
            a_thread_panicked: AtomicBool::new(false),
        }),
        // ...
    };

    let result = catch_unwind(AssertUnwindSafe(|| f(&scope)));

    // Wait until all the threads are finished.
    while scope.data.num_running_threads.load(Ordering::Acquire) != 0 {
        park();
    }
    // ...
}
```

## (1) Who calls `unpark()`?

```rust ignore
impl ScopeData {
    pub(super) fn decrement_num_running_threads(&self, panic: bool) {
        // ...
        // fetch_sub returns the previous value
        if self.num_running_threads.fetch_sub(1, Ordering::Release) == 1 {
            self.main_thread.unpark();
        }
    }
}
```

## (2) Who calls `unpark()`?

```rust ignore
impl<'scope, T> Drop for Packet<'scope, T> {
    fn drop(&mut self) {
        // ...
        if let Some(scope) = &self.scope {
            scope.decrement_num_running_threads(/* ... */);
        }
    }
}
```

> 

Reviewer: Links to stdlib source code here would be neat.

`Packet` is a mechanism to pass panics and results from a thread closure to the parent thread.

## `thread::scope()`

> * function completion is controlled by a single atomic counter in an `Arc`

Reviewer: ```suggestion
* function completion is monitored by a single atomic counter in an `Arc`
```

Reviewer: actually, scratch this. I read this now as "completion of the `scope()` function", and the word *controlled* is therefore appropriate.
